nefc refactor

miguelangel-dev opened the pull request:
## Related issues
 - #112
 - #122 

## Description
In this PR, we have added a new module `NefCompiler` using `NefRender` to get the whole `playgrounds > pages` and compile them. Also, this PR contains the CLI `Compiler/main.swift` to use the nef API.

### Modules
- Compiler
- NefCompiler

miguelangel-dev committed:
structure for Nef Compiler

miguelangel-dev committed:
NefCompiler main-logic and algebras + SPM update

miguelangel-dev committed:
extract scheme name given a xcworkspace

miguelangel-dev committed:
build project

miguelangel-dev committed:
compile project with dependencies

miguelangel-dev committed:
get the whole resources/dependencies for compiling a page

miguelangel-dev committed:
add compile playground (from URL) action to CompilerAPI

miguelangel-dev committed:
add temporal file operation to FileSystem Algebra

miguelangel-dev committed:
compile swift page from a Playground together with sources

miguelangel-dev committed:
add logger for compiler

miguelangel-dev committed:
adding details log in NefCompiler

miguelangel-dev committed:
api docs for compiler API

miguelangel-dev committed:
Merge branch 'develop' into nefc-refactor

miguelangel-dev committed:
merge develop into nefc-refactor branch

miguelangel-dev committed:
add helpers for running nef API

truizlop commented on project/Component/NefRender/Errors/MarkdownError.swift:
```diff
@@ -3,29 +3,35 @@
-        case .content:
-            return "can not render content"
+        case .content(let info):
+            return "Can not render content\(info.isEmpty ? "" : ": \(info)")"
```
```suggestion
            return "Cannot render content\(info.isEmpty ? "" : ": \(info)")"
```

truizlop commented on project/Component/NefRender/Errors/MarkdownError.swift:
```diff
@@ -3,29 +3,35 @@
+            return "Can not render content\(info.isEmpty ? "" : ": \(info)")"
         case .page(let page):
-            return "can not render page '\(page.path)'"
+            return "Can not render page '\(page.path)'"
```
```suggestion
            return "Cannot render page '\(page.path)'"
```

truizlop commented on project/Component/NefRender/Errors/MarkdownError.swift:
```diff
@@ -3,29 +3,35 @@
+            return "Can not render page '\(page.path)'"
         case .playground(let playground):
-            return "can not render playground '\(playground.path)'"
+            return "Can not render playground '\(playground.path)'"
```
```suggestion
            return "Cannot render playground '\(playground.path)'"
```

truizlop commented on project/Component/NefRender/Errors/MarkdownError.swift:
```diff
@@ -3,29 +3,35 @@
-            return "can not render playground '\(playground.path)'"
+            return "Can not render playground '\(playground.path)'"
+        case .workspace(let workspace, let info):
+            return "Can not render workspace '\(workspace.path)'\(info.isEmpty ? "" : ". \n  Reason: \(info.description.firstLowercased)")"
```
```suggestion
            return "Cannot render workspace '\(workspace.path)'\(info.isEmpty ? "" : ". \n  Reason: \(info.description.firstLowercased)")"
```

truizlop commented on project/Component/NefRender/Errors/MarkdownError.swift:
```diff
@@ -3,29 +3,35 @@
+            return "Can not render workspace '\(workspace.path)'\(info.isEmpty ? "" : ". \n  Reason: \(info.description.firstLowercased)")"
         case .playgrounds:
-            return "can not render playgrounds"
+            return "Can not render playgrounds"
```
```suggestion
            return "Cannot render playgrounds"
```

truizlop commented on project/Component/nef/Models/Error.swift:
```diff
@@ -2,10 +2,30 @@
+    public var description: String {
+        switch self {
+        case .compiler(let info):
+            return info.isEmpty ? "Failed the compiler." : info
```
```suggestion
            return info.isEmpty ? "Compiler failure." : info
```

truizlop commented on project/Component/nef/Models/Error.swift:
```diff
@@ -2,10 +2,30 @@
+        case .compiler(let info):
+            return info.isEmpty ? "Failed the compiler." : info
+        case .markdown(let info):
+            return info.isEmpty ? "Failed the markdown." : info
```
```suggestion
            return info.isEmpty ? "Markdown failure." : info
```

truizlop commented on project/Component/nef/Models/Error.swift:
```diff
@@ -2,10 +2,30 @@
+        case .markdown(let info):
+            return info.isEmpty ? "Failed the markdown." : info
+        case .jekyll(let info):
+            return info.isEmpty ? "Failed the jekyll." : info
```
```suggestion
            return info.isEmpty ? "Jekyll failure." : info
```

truizlop commented on project/Component/nef/Models/Error.swift:
```diff
@@ -2,10 +2,30 @@
+        case .jekyll(let info):
+            return info.isEmpty ? "Failed the jekyll." : info
+        case .carbon(let info):
+            return info.isEmpty ? "Failed the carbon." : info
```
```suggestion
            return info.isEmpty ? "Carbon failure." : info
```

truizlop commented on project/Component/nef/Models/Error.swift:
```diff
@@ -2,10 +2,30 @@
+        case .invalidSnapshot(let info):
+            return info.isEmpty ? "Invalid snapshot." : info
+        case .swiftPlaygrond(let info):
+            return info.isEmpty ? "Failed Swift Playground generator." : info
```
```suggestion
            return info.isEmpty ? "Swift Playground generator failure." : info
```

truizlop commented on project/UI/CLIKit/Console.swift:
```diff
@@ -179,7 +181,7 @@ public struct Console {
                 return "do not received the whole required arguments."+" Use".bold+" --help, --h".cyan
             case .render(let info):
-                return "fail the render\(info.isEmpty ? "" : " (".lightGray+"\(info)".lightRed+")".lightGray)."
+                return info.isEmpty ? "" : "Fail the render: \(info.lightRed)"
```
```suggestion
                return info.isEmpty ? "" : "Render failure: \(info.lightRed)"
```

truizlop commented on project/Component/NefCompiler/Errors/CompilerShellError.swift:
```diff
@@ -0,0 +1,20 @@
+            return "command not found '\(command)' \(information.isEmpty ? "." : " (\(information))")"
+            
+        case .failed(let command, let information):
+            return "failed command '\(command)' \(information)"
```
```suggestion
            return "command '\(command)' failed: \(information)"
```

truizlop commented on project/Component/NefCompiler/Errors/CompilerShellError.swift:
```diff
@@ -0,0 +1,20 @@
+    public var description: String {
+        switch self {
+        case .notFound(let command, let information):
+            return "command not found '\(command)' \(information.isEmpty ? "." : " (\(information))")"
```
```suggestion
            return "command '\(command)' not found: \(information.isEmpty ? "." : " (\(information))")"
```

miguelangel-dev committed:
comments in PR

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/NefRender/Errors/MarkdownError.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/NefRender/Errors/MarkdownError.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/NefRender/Errors/MarkdownError.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/NefRender/Errors/MarkdownError.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/nef/Models/Error.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/UI/CLIKit/Console.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/NefCompiler/Errors/CompilerShellError.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/NefCompiler/Errors/CompilerShellError.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/nef/Models/Error.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/nef/Models/Error.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/nef/Models/Error.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>

Update project/Component/nef/Models/Error.swift

Co-Authored-By: Tomás Ruiz-López <truizlop@gmail.com>
